1_fetch/src/fetch_wqp_data.R

```diff
@@ -47,12 +65,63 @@ add_download_groups <- function(sitecounts_df, max_sites = 500, max_results = 25
                                               maxgroupsize = max_sites),
                download_grp = paste0(grid_id,"_",task_num))
     }) %>%
-    mutate(site_n = row_number()) %>%
-    select(site_id, results_count, site_n, download_grp)
+    mutate(pull_id = TRUE)
   
-  return(sitecounts_grouped)
+  # Assign a separate task number and download group for all sites with 
+  # bad identifiers within a grid
+  sitecounts_grouped_bad_ids <- sitecounts_bad_ids %>%
+    group_by(grid_id) %>%
```

Comment:
There are examples of grids like this for the 'alternate' wqp args at the top of this PR. I first coded this up as each bad id gets its own query, but then I wondered if that was less efficient than querying all the 'bad id' data within a grid together, and then subsetting the downloaded data by the identifiers we actually wanted. However, I think it's worth revisiting this because depending on the grid size a user uses we could end up querying much larger bounding boxes than I anticipated, which would not be very efficient! Here's an example of multiple bad id's in the same grid, so you can see we're downloading quite a few extra sites than we were necessarily looking for. 

```r
> tar_load(p1_site_counts)
> tar_load(p1_char_names)
> 
> # Part of the code that groups sites
> sitecounts_bad_ids <- p1_site_counts %>%
+     filter(grid_id == "11744",
+            grepl("/", MonitoringLocationIdentifier)) %>%
+     rename(site_id = MonitoringLocationIdentifier)
> site_counts_grouped <- sitecounts_bad_ids %>%
+     group_by(grid_id) %>%
+     mutate(task_num = max(sitecounts_grouped_good_ids$task_num) + cur_group_id(),
+            download_grp = paste0(grid_id,"_",task_num),
+            pull_by_id = FALSE) %>%
+     ungroup()
> 
> # Pull the data for these 'bad' sites
> bbox <- create_site_bbox(site_counts_grouped)
> wqp_args_all <- list(sampleMedia = c("Water","water"),
+                      siteType = "Lake, Reservoir, Impoundment",
+                      # return sites with at least one data record
+                      minresults = 1, 
+                      startDateLo = "2000-01-01",
+                      startDateHi = "2020-12-31",
+                      bBox = bbox, 
+                      characteristicName = c(p1_char_names))
> wqp_data <- retry::retry(dataRetrieval::readWQPdata(wqp_args_all),
+                          when = "Error:", 
+                          max_tries = max_tries)
 0s>                                                                                                                                            0s
> # Do we pick up 'extra' sites? (Yes...)
> wqp_data %>%
+     filter(!MonitoringLocationIdentifier %in% site_counts_grouped$site_id) %>%
+     pull(MonitoringLocationIdentifier) %>% unique()
 [1] "PCCB_WQX-977163"           "USGS-05485511"             "USGS-05485512"             "USGS-05485513"            
 [5] "USGS-05485514"             "21IOWA_WQX-22770001"       "21IOWA_WQX-22770005"       "21IOWA_WQX-22630002"      
 [9] "21IOWA_WQX-22770002"       "21IOWA_WQX-22770003"       "21IOWA_WQX-22770006"       "21IOWA_WQX-14000250"      
[13] "21IOWA-22770005"           "21IOWA-22770002"           "21IOWA-22770001"           "21IOWA-22630002"          
[17] "21IOWA-22770003"           "NALMS-C41.7044,-93.6817"   "NALMS-C41.59831,-93.60861" "NALMS-F460560"            
[21] "NARS_WQX-NLA06608-2036"    "NARS_WQX-NLA12_IA-103"    
>
```

Reply:
I see you implemented one new query per bad site. That is the way I think I am leaning, too, so let's keep that

Reply:
I agree - I think the two approaches could be comparable for small grids, but I can also envision cases where the resulting bbox results in a much larger query than we bargained for. One query per bad site seems safer to me.